### PR TITLE
Implement dynamic linking support for ASan/LSan

### PR DESCRIPTION
**Pull request by quantum5**
Implement replacements for `__data_end`, `__heap_base` for when compiling ASan in PIC mode.

Things to do before this support is complete:

* [ ] LSan needs to scan global variables of shared libraries. This involves enumerating all libraries at runtime, and finding their data sections.
* [ ] PC value assignment for shared libraries. This will allow us to identify call sites in dynamic libraries.
* [ ] Source map support for dynamic libraries.

### CODE REVIEW

File: system/lib/compiler-rt/lib/lsan/lsan_common_emscripten.cc [hunk 1]
```diff
@@ -77,15 +77,14 @@ void InitializePlatformSpecificModules() {
...
-
 // Scans global variables for heap pointers.
 void ProcessGlobalRegions(Frontier *frontier) {
   if (!flags()->use_globals) return;
-  ScanGlobalRange(0, (uptr) &__data_end, frontier);
+  ScanGlobalRange(EM_ASM_INT({
+    return GLOBAL_BASE;
+  }), EM_ASM_INT({
+    return STACK_TOP;
+  }), frontier);
```
Comment (tlively) [inline system/lib/compiler-rt/lib/lsan/lsan_common_emscripten.cc]: How is this related to PIC? Does the behavior change in the non-PIC case? It's unfortunate to have a tight dependency on Emscripten in code that we would ideally upstream at some point.

Comment (quantum5) [reply system/lib/compiler-rt/lib/lsan/lsan_common_emscripten.cc]: This is the emscripten-specific implementation for this function. The behaviour does not change in the non-PIC case here, as we are simply using alternatives to `__data_end`.

**Comment by sbc100** [comment]
The deletion of the `incoming` branch triggered the automatic closing of this branch.  My apologies.  This was an unexpected side effect.

Re-opening and re-targeting to master.

**Comment by sbc100** [comment]
This is still a goal.  I'm not sure if this change is relevant but perhaps it can still be rebased?
